coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,241 @@
+func WithFlushChannel(c chan int) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithFlushChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only and will panic if used outside of tests.
+func WithTickChannel(c chan time.Time) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithTickChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.m.Add(workspaceID)
+}
+
+// flush updates last_used_at of all current workspace IDs.
+// If this is held while a previous flush is in progress, it will
+// deadlock until the previous flush has completed.
+func (wut *Tracker) flush(now time.Time) {
+	// Copy our current set of IDs
+	ids := wut.m.UniqueAndClear()
+	count := len(ids)
+	if wut.flushCh != nil { // only used for testing
+		defer func() {
+			wut.flushCh <- count
+		}()
+	}
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+
+	// Set a short-ish timeout for this. We don't want to hang forever.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// nolint: gocritic // system function
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	wut.flushLock.Lock()
+	defer wut.flushLock.Unlock()
+	if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
+		LastUsedAt: now,
+		IDs:        ids,
+	}); err != nil {
+		// A single failure to flush is likely not a huge problem. If the workspace is still connected at
+		// the next iteration, either another coderd instance will likely have this data or the CLI
+		// will tell us again that the workspace is in use.
+		wut.flushErrors++
+		if wut.flushErrors > 1 {
+			wut.log.Error(ctx, "multiple failures updating workspaces last_used_at", slog.F("count", count), slog.F("consecutive_errors", wut.flushErrors), slog.Error(err))
+			// TODO: if this keeps failing, it indicates a fundamental problem with the database connection.
+			// How to surface it correctly to admins besides just screaming into the logs?
+		} else {
+			wut.log.Warn(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
+		}
+		return
+	}
+	wut.flushErrors = 0
+	wut.log.Info(ctx, "updated workspaces last_used_at", slog.F("count", count), slog.F("now", now))
+}
+
+// Loop periodically flushes every tick.
+// If Loop is called after Close, it will panic. Don't do this.
+func (wut *Tracker) Loop() {
+	// Calling Loop after Close() is an error.
+	select {
+	case <-wut.doneCh:
+		panic("developer error: Loop called after Close")
```

Comment:
If we need to do this instead of just returning, I'd prefer returning an error here, wouldn't want this to happen for a production instance after some changes to the code.

Feels like this mostly exists for tests (otherwise could be launched from `New`), so more reason not to add a possibility of panic.

Reply:
If we return an error, then we have two options:
1) Exit with the error, which brings us back to potentially impacting prod, but just in a nicer-looking way.
2) Just log the error and do nothing, which means the error will likely pass unnoticed until folks wonder why their workspace usage isn't being tracked.

I don't think 2) is a valid option. We're not guaranteed to catch it in tests especially because of the possibility of ignoring errors in `slogtest`.

There's an alternative possibility 1a) where we check the error and recreate the tracker if non-nil. WDYT?

Reply:
I forgot about option c) - have `New` start the loop and unexport the function.
Now it's un-possible!